sql/procedures/parse.go

```diff
@@ -0,0 +1,368 @@
+		selectCond := &ast.Select{
+			SelectExprs: ast.SelectExprs{
+				&ast.AliasedExpr{
+					Expr: whileCond,
+				},
+			},
+		}
+		whileOp := &InterpreterOperation{
+			OpCode:      OpCode_If,
+			PrimaryData: selectCond,
+		}
+		*ops = append(*ops, whileOp)
+
+		for _, whileStmt := range s.Statements {
+			if err := ConvertStmt(ops, stack, whileStmt); err != nil {
+				return err
+			}
+		}
+		gotoOp := &InterpreterOperation{
+			OpCode: OpCode_Goto,
+			Index:  loopStart,
+		}
+		*ops = append(*ops, gotoOp)
+		whileOp.Index = len(*ops) // end of while block
+		resolveGoToIndexes(ops, s.Label, loopStart, whileOp.Index, loopStart, whileOp.Index)
+
+	case *ast.Repeat:
+		// repeat statements always run at least once
+		onceStart := len(*ops)
+		for _, repeatStmt := range s.Statements {
+			if err := ConvertStmt(ops, stack, repeatStmt); err != nil {
+				return err
+			}
+		}
+
+		loopStart := len(*ops)
+		if s.Label != "" {
+			stack.NewLabel(s.Label, loopStart)
+		}
+		repeatCond := &ast.NotExpr{Expr: s.Condition}
+		selectCond := &ast.Select{
+			SelectExprs: ast.SelectExprs{
+				&ast.AliasedExpr{
+					Expr: repeatCond,
+				},
+			},
+		}
+		repeatOp := &InterpreterOperation{
+			OpCode:      OpCode_If,
+			PrimaryData: selectCond,
+		}
+		*ops = append(*ops, repeatOp)
+
+		for _, repeatStmt := range s.Statements {
+			if err := ConvertStmt(ops, stack, repeatStmt); err != nil {
+				return err
+			}
+		}
+
+		gotoOp := &InterpreterOperation{
+			OpCode: OpCode_Goto,
+			Index:  loopStart,
+		}
+		*ops = append(*ops, gotoOp)
+		repeatOp.Index = len(*ops) // end of repeat block
+		resolveGoToIndexes(ops, s.Label, onceStart, repeatOp.Index, loopStart, repeatOp.Index)
+
+	case *ast.Loop:
+		loopStart := len(*ops)
+		if s.Label != "" {
+			stack.NewLabel(s.Label, loopStart)
+		}
+		for _, loopStmt := range s.Statements {
+			if err := ConvertStmt(ops, stack, loopStmt); err != nil {
+				return err
+			}
+		}
+		gotoOp := &InterpreterOperation{
+			OpCode: OpCode_Goto,
+			Target: s.Label,
+			Index:  loopStart,
+		}
+		*ops = append(*ops, gotoOp)
+		loopEnd := len(*ops)
+		resolveGoToIndexes(ops, s.Label, loopStart, loopEnd, loopStart, loopEnd)
+
+	case *ast.Iterate:
+		iterateOp := &InterpreterOperation{
+			OpCode: OpCode_Goto,
+			Target: s.Label,
+			Index:  stack.GetLabel(s.Label), // possible this is -1, which will get resolved later
+		}
+		*ops = append(*ops, iterateOp)
+
+	case *ast.Leave:
+		leaveOp := &InterpreterOperation{
+			OpCode: OpCode_Goto,
+			Target: s.Label,
+			Index:  -2, // -2 indicates that this is a leave statement with unknown target index
+		}
+		*ops = append(*ops, leaveOp)
+
+	default:
+		executeOp := &InterpreterOperation{
+			OpCode:      OpCode_Execute,
+			PrimaryData: s,
+		}
+		*ops = append(*ops, executeOp)
+	}
+
+	return nil
+}
+
+// Parse parses the given CREATE FUNCTION string (which must be the entire string, not just the body) into a Block
+// containing the contents of the body.
```

Comment:
Comment from Doltgres, this isn't a string in GMS. Also parses to a set of operations rather than a block.